llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -1195,6 +1195,25 @@ translateSpirvGlobalUses(LoadInst *LI, StringRef SpirvGlobalName,
     return;
   }
 
+  // As an optimization of accesses of the first element for vector SPIRV
+  // globals sometimes the load will return a scalar and the uses can be of any
+  // pattern. In this case, generate GenX calls to access the first element and
+  // update the use to instead use the GenX call result rather than the load
+  // result.
+  if (!LI->getType()->isVectorTy()) {
+    // Copy users to seperate container for safe modification
+    // during iteration.
+    SmallVector<User *> Users(LI->users().begin(), LI->users().end());
```

Comment:
This SmallVector seems redundant, but I may be wrong. 
Nick, can you imagine a case when the loop below would work incorrectly without it?

Reply:
Yes, I added that specifically because we got an assert without it. Note we do the same thing here https://github.com/intel/llvm/blob/sycl/llvm/lib/SYCLLowerIR/ESIMD/ESIMDOptimizeVecArgCallConv.cpp#L483

Reply:
Yeah, replacing users of an instruction on-the-fly, while iterating over them is a bad idea. BTW, `SmallVector` should support constructing it from a range as well: [doxygen link](https://llvm.org/doxygen/classllvm_1_1SmallVector.html#a6e4373e43164f1022d8b63a149f5ae5a)

```suggestion
    SmallVector<User *> Users(LI->users());
```

Reply:
ah i missed that ctor, thanks

Reply:
@sarnex - shouldn't the same idea be used then in the next old loop (following the one you added) ?

Reply:
@v-klochkov i think there's a difference. in my loop, we are adding/removing the users of LI while iterating over LI users. In the other loop, we are adding/removing the users of LU while iterating over LI users. The second loop isn't changing the users of the thing being iterated over. I think that's the difference.